### DISCUSSION TITLE
How to make the server not reload automatically, ever?

### DISCUSSION THREAD

**Discussion by abhishekrai43** [zauberzeug/nicegui#2339] [Q&A]
### Description

I cannot get watchfiles to exclude my current dir I have tried various glob patterns, Also when there is an error it again reloads, so For e.g, Autogen is taking human input, it somehow treats it as a some error/reset, it will reload and the whole conversation is gone 
`ui.run(title='Chat with Autogen Assistant',host='localhost',reload=False)` doesn't work. 
Neither do these
```
Assistant',host='localhost',uvicorn_reload_excludes='.groupchat/**/*')
uvicorn_reload_excludes='.*,'
uvicorn_reload_excludes='**/*
```'

**Comment by falkoschindler** [comment]
Hi @abhishekrai43!
Have you restarted the application after changing these parameters?

**Reply by abhishekrai43** [reply]
Number of times, that's all what I have been trying infact. Changing the glob pattern etc and restarting in hopes it won't reload

**Reply by falkoschindler** [reply]
Does a simple hello-world app has the same problem if you try to run it with `reload=False`?
If yes, can you check if a plain FastAPI app also keeps reloading without the `--reload` parameter?
If no, can you boil it down to a minimal reproducible example?
Besides that, what are your NiceGUI, FastAPI, uvicorn and watchfiles versions?

**Reply by abhishekrai43** [reply]
I think the best I can boil it down to is to related to Autogen, the filewatcher thing is still manageable. 
print a list in python

--------------------------------------------------------------------------------
Provide feedback to chat_manager. Press enter to skip and use auto-reply, or type 'exit' to end the conversation: 

This "human input" (when the mode is 'ALWAYS', added code below) triggers the reload of the server and I cannot proceed. I am stuck here because this fails everything, so some hint/ reference on how to approach towards solving this would do it for me.

**Reply by abhishekrai43** [reply]
```py
def check_termination_and_human_reply(
        self,
        messages: Optional[List[Dict]] = None,
        sender: Optional[Agent] = None,
        config: Optional[Any] = None,
    ) -> Tuple[bool, Union[str, None]]:
        """Check if the conversation should be terminated, and if human reply is provided.

        This method checks for conditions that require the conversation to be terminated, such as reaching
        a maximum number of consecutive auto-replies or encountering a termination message. Additionally,
        it prompts for and processes human input based on the configured human input mode, which can be
        'ALWAYS', 'NEVER', or 'TERMINATE'. The method also manages the consecutive auto-reply counter
        for the conversation and prints relevant messages based on the human input received.

        Args:
            - messages (Optional[List[Dict]]): A list of message dictionaries, representing the conversation history.
            - sender (Optional[Agent]): The agent object representing the sender of the message.
            - config (Optional[Any]): Configuration object, defaults to the current instance if not provided.

        Returns:
            - Tuple[bool, Union[str, Dict, None]]: A tuple containing a boolean indicating if the conversation
            should be terminated, and a human reply which can be a string, a dictionary, or None.
        """
        # Function implementation...

        if config is None:
            config = self
        if messages is None:
            messages = self._oai_messages[sender]
        message = messages[-1]
        reply = ""
        no_human_input_msg = ""
        if self.human_input_mode == "ALWAYS":
            reply = self.get_human_input(
                f"Provide feedback to {sender.name}. Press enter to skip and use auto-reply, or type 'exit' to end the conversation: "
            )
            no_human_input_msg = "NO HUMAN INPUT RECEIVED." if not reply else ""
            # if the human input is empty, and the message is a termination message, then we will terminate the conversation
            reply = reply if reply or not self._is_termination_msg(message) else "exit"
        else:
            if self._consecutive_auto_reply_counter[sender] >= self._max_consecutive_auto_reply_dict[sender]:
                if self.human_input_mode == "NEVER":
                    reply = "exit"
                else:
                    # self.human_input_mode == "TERMINATE":
                    terminate = self._is_termination_msg(message)
                    reply = self.get_human_input(
                        f"Please give feedback to {sender.name}. Press enter or type 'exit' to stop the conversation: "
                        if terminate
                        else f"Please give feedback to {sender.name}. Press enter to skip and use auto-reply, or type 'exit' to stop the conversation: "
                    )
                    no_human_input_msg = "NO HUMAN INPUT RECEIVED." if not reply else ""
                    # if the human input is empty, and the message is a termination message, then we will terminate the conversation
                    reply = reply if reply or not terminate else "exit"
            elif self._is_termination_msg(message):
                if self.human_input_mode == "NEVER":
                    reply = "exit"
                else:
                    # self.human_input_mode == "TERMINATE":
                    reply = self.get_human_input(
                        f"Please give feedback to {sender.name}. Press enter or type 'exit' to stop the conversation: "
                    )
                    no_human_input_msg = "NO HUMAN INPUT RECEIVED." if not reply else ""
                    # if the human input is empty, and the message is a termination message, then we will terminate the conversation
                    reply = reply or "exit"

        # print the no_human_input_msg
        if no_human_input_msg:
            print(colored(f"\n>>>>>>>> {no_human_input_msg}", "red"), flush=True)

        # stop the conversation
        if reply == "exit":
            # reset the consecutive_auto_reply_counter
            self._consecutive_auto_reply_counter[sender] = 0
            return True, None

        # send the human reply
        if reply or self._max_consecutive_auto_reply_dict[sender] == 0:
            # reset the consecutive_auto_reply_counter
            self._consecutive_auto_reply_counter[sender] = 0
            return True, reply

        # increment the consecutive_auto_reply_counter
        self._consecutive_auto_reply_counter[sender] += 1
        if self.human_input_mode != "NEVER":
            print(colored("\n>>>>>>>> USING AUTO REPLY...", "red"), flush=True)

        return False, None
```
Not sure what about this message would make the server reload?

**Reply by falkoschindler** [reply]
This looks like a small piece from a much larger codebase. Without a [minimal reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example) it's very hard and time-consuming to help.
You also didn't answer [my questions](https://github.com/zauberzeug/nicegui/discussions/2339#discussioncomment-8085741).

**Reply by abhishekrai43** [reply]
Agreed, Unfortunately, I have to either get this to work or find alternatives, cannot spend time on anything else. I shall close this then

**Comment by merlinarer** [comment]
Hello @abhishekrai43 @falkoschindler , have you find the reason? I also meet this issue now in my chatbot project. The page sometimes reload during chatting just as reload=True even I already set reload=False.

**Reply by falkoschindler** [reply]
@merlinarer No, there hasn't been any new progress on this discussion. But I actually wonder if it's really the server that is reloading (which would be controlled by `reload=False`) or if it's just the browser that is reloading the page. The latter could be caused by too [much load on the main thread](https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-my-long-running-function-blocking-ui-updates). Maybe you can give more insight on what is happening.

**Reply by merlinarer** [reply]
Hi @falkoschindler, I paste here my reproducible code:
```py
#!/usr/bin/env python3
from typing import List, Tuple

from langchain.chains import ConversationChain
from langchain.chat_models import ChatOpenAI
import openai
from nicegui import context, ui
from langchain_core.messages import (
    AIMessageChunk,
    FunctionMessageChunk,
    HumanMessageChunk,
    SystemMessageChunk,
    ToolMessageChunk,
)
import httpx
OPENAI_API_KEY = 'sk-----'  # TODO: set your OpenAI API key here
openai.proxy = "http://localhost:4780"

@ui.page('/')
def main():
    llm=ChatOpenAI(model_name='gpt-3.5-turbo', openai_api_key=OPENAI_API_KEY )
    messages: List[Tuple[str, str]] = []
    thinking: bool = False

    @ui.refreshable
    def chat_messages() -> None:
        for name, text in messages:
            ui.chat_message(text=text, name=name, sent=name == 'You')
        if thinking:
            ui.spinner(size='3rem').classes('self-center')
        if context.get_client().has_socket_connection:
            ui.run_javascript('window.scrollTo(0, document.body.scrollHeight)')

    async def send() -> None:
        nonlocal thinking
        message = text.value
        messages.append(('You', text.value))
        thinking = True
        text.value = ''
        chat_messages.refresh()
        messages.append(['Bot', ""])
        async for chunk in llm.astream(message):
            messages[-1][-1] += chunk.content
            thinking = False
            chat_messages.refresh()
   
    anchor_style = r'a:link, a:visited {color: inherit !important; text-decoration: none; font-weight: 500}'
    ui.add_head_html(f'<style>{anchor_style}</style>')

    # the queries below are used to expand the contend down to the footer (content can then use flex-grow to expand)
    ui.query('.q-page').classes('flex')
    ui.query('.nicegui-content').classes('w-full')

    with ui.tabs().classes('w-full') as tabs:
        chat_tab = ui.tab('Chat')
        logs_tab = ui.tab('Logs')
    with ui.tab_panels(tabs, value=chat_tab).classes('w-full max-w-2xl mx-auto flex-grow items-stretch'):
        with ui.tab_panel(chat_tab).classes('items-stretch'):
            chat_messages()
        with ui.tab_panel(logs_tab):
            log = ui.log().classes('w-full h-full')

    with ui.footer().classes('bg-white'), ui.column().classes('w-full max-w-3xl mx-auto my-6'):
        with ui.row().classes('w-full no-wrap items-center'):
            placeholder = 'message' if OPENAI_API_KEY != 'not-set' else \
                'Please provide your OPENAI key in the Python script first!'
            text = ui.input(placeholder=placeholder).props('rounded outlined input-class=mx-3') \
                .classes('w-full self-center').on('keydown.enter', send)
        ui.markdown('simple chat app built with [NiceGUI](https://nicegui.io)') \
            .classes('text-xs self-end mr-8 m-[-1em] text-primary')

ui.run(title='Chat with GPT-3 (example)', reload=False)
```

**Reply by merlinarer** [reply]
With this code, when chatting with long context several times, reloading will happen. It seems something wrong with refresh.
```
49304
Traceback (most recent call last):
  File "C:\Users\1\AppData\Local\miniconda3\lib\site-packages\nicegui\events.py", line 406, in wait_for_result
    await result
  File "D:\Proj\test_reload.py", line 46, in send
    chat_messages.refresh()
  File "C:\Users\1\AppData\Local\miniconda3\lib\site-packages\nicegui\functions\refreshable.py", line 79, in refresh
    attribute(*args, **kwargs)
  File "C:\Users\1\AppData\Local\miniconda3\lib\site-packages\nicegui\functions\refreshable.py", line 100, in refresh
    target.container.clear()
  File "C:\Users\1\AppData\Local\miniconda3\lib\site-packages\nicegui\element.py", line 460, in clear
    self.client.remove_elements(descendants)
  File "C:\Users\1\AppData\Local\miniconda3\lib\site-packages\nicegui\client.py", line 298, in remove_elements
    del self.elements[element_id]
KeyError: 49304
```

**Reply by merlinarer** [reply]
here is a video for overview

https://github.com/zauberzeug/nicegui/assets/71754029/ac8b68de-2e14-4188-ab33-9b6256074a39

**Reply by merlinarer** [reply]
finally uploaded, I use a gpt3.5 openaikey for test, and I will share with you for test if needed

**Reply by falkoschindler** [reply]
Thanks, @merlinarer!
The code is pretty similar to our original ["Chat with AI" example](https://github.com/zauberzeug/nicegui/tree/main/examples/chat_with_ai). Can you reproduce the problem with the original code as well?

**Reply by merlinarer** [reply]
Not similar, but the same~

**Reply by abhishekrai43** [reply]
@merlinarer , Ideally `Reload=False` should suffice. In case that is not working, see if the error generated is changing any files in the directory..any changes to any files will trigger reload

**Reply by merlinarer** [reply]
hi @abhishekrai43 `Reload=False` do work when I change the codes or files and it seem something in refresh trigger reload.

**Reply by falkoschindler** [reply]
@merlinarer No, the code is not the same (e.g. `llm.astream` vs. `llm.arun`). I'm just not sure if the diff is relevant.

@abhishekrai43 I'm not sure if the server is actually reloading. Can anyone confirm that? @merlinarer Do you see something like "WatchFiles detected changes in ... Reloading..." on the command line?

**Reply by abhishekrai43** [reply]
@falkoschindler Yes, that is what I was facing. That got fixed with `reload=False` is what I meant to say.

**Reply by merlinarer** [reply]
oh, yeah, I change to `llm.astream` just for streaming output. I also try my local model server with `httpx stream`, and the same reloading happens. There is no anything like "WatchFiles detected changes in ... Reloading.." on the command line.

**Reply by falkoschindler** [reply]
@merlinarer So the server is _not_ reloading, which means you are facing a different problem than @abhishekrai43. Again: Please try the original "Chat with AI" example. I'm not too familiar with the `llm` API, but maybe `llm.astream` is blocking for too long.

**Reply by merlinarer** [reply]
emm, I'm not sure what is the definition of "the server is reloading". When this problem happens, all the element the page will recreated again just as first open this page.

**Comment by abhishekrai43** [comment]
> @merlinarer No, the code is not the same (e.g. `llm.astream` vs. `llm.arun`). I'm just not sure if the diff is relevant.
> 
> @abhishekrai43 I'm not sure if the server is actually reloading. Can anyone confirm that? @merlinarer Do you see something like "WatchFiles detected changes in ... Reloading..." on the command line?

@merlinarer is right when he says 'Server reloads'..The server itself reloads...so it will wipe out everything and start refresh.

**Comment by merlinarer** [comment]
@falkoschindler  I tried the original "Chat with AI" example just now，but the `llm = ConversationChain(llm=ChatOpenAI(model_name='gpt-3.5-turbo', openai_api_key=OPENAI_API_KEY)) llm.arun` seems automatically add history to openai，which leads to accepting much shorter input, and I was unable to reproduce long context.

**Comment by falkoschindler** [comment]
Let me try to summarize:

- The original question from @abhishekrai43 was about the server (the Python process) restarting even though auto-reloading was disabled (`ui.run(..., reload=False)`). This problem got [closed](https://github.com/zauberzeug/nicegui/discussions/2339#discussioncomment-8092386) due to the lack of a reproduction.
- @merlinarer's Problem doesn't seem to be caused by a server reload ("There is no anything like "WatchFiles detected changes in ... Reloading.." on the command line."). Instead it could be a bug in the UI causing it to clear all content, or - because something is blocking the main thread on the server for too long - the browser reloading the page which caused it to be created from scratch. But I still can't reproduce it and I don't really know what's happening.

I suggest to conclude this discussion and create a new Q&A for individual follow-up problems. This would help focussing on specific scenarios and avoid mixing things up.
